webidls/Pipeline.webidl

```diff
@@ -0,0 +1,176 @@
+}
+
+enum SaveRenderingStrategy {
+  "default",
+  "displaylist",
+  "texture",
+}
+
+interface Pipeline {
+
+  Promise<void> setPipelineHandler(PipelineHandler handler);
+
+  readonly attribute USVString url; // Happens during redirects for example. FIXME: should that be finalURL ? How often would that change?
+  readonly attribute DOMString title;
+  readonly attribute unsigned short HTTPResponse;
+
+  // Used to replace mozbrowserconnected, mozbrowserloadstart, mozbrowserloadend
+  // Use performance for time stamps.
+  readonly attribute PipelineState? pipelineState;
+  readonly attribute Performance performance; // See Performance.webidl and PerformanceTiming.webidl // FIXME: should probably be getPerformance(). Necessary to implement https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/history/HistoryItem
+  readonly attribute PipelineError pipelineError; // pipeline-state-changed -> pipelineState == "error" || "crash". // FIXME: is "crash" necessary? // FIXME: so generic…
+
+  readonly attribute FrozenArray<USVString> icons;
+  readonly attribute FrozenArray<MetaTag> metas;
+  readonly attribute String? hoveredLink;
+  readonly attribute ConnectionSecurity connectionSecurity;
+
+  readonly attribute boolean isFrozen;
+  readonly attribute boolean isVisible;
+  readonly attribute unsigned float devicePixelRatio;
+
+  Promise<unsigned float> setDevicePixelRatio(unsigned float ratio); // Will fail for non-top level pipelines
+
+  // Prerendering a document. HTTP + layout, no JS. Lot of things to consider. See:
+  // https://bugzilla.mozilla.org/show_bug.cgi?id=730101
+  // Will be rendered once Browser.navigateToPipeline() is called.
+  // Created via Browser.createPrerenderingPipeline()
+  readonly attribute boolean isPrerender;
+  Promise<void> cancelPrerenderAndRequestDestruction();
+
+  void stopLoading();
+
+  // Will destroy that pipeline and create a new one
+  void reload();
+  void clearCacheAndReload();
+
+  Promise<LoadData> buildLoadData();
+
+  // We want to be able to render frozen pipeline, so we need
+  // a way to save the rendering.
+  // FIXME: do we really want to let the client handle that?
+  readonly attribute SaveRenderingStrategy saveRenderingStrategy;
+  Promise<void> setSaveRenderingStrategy(SaveRenderingStrategy);
+
+  Promise<void> insertCSS(DOMString code);
+
+  Promise<Object> evaluateScript(DOMString script, boolean onlyForFrameScript);
+  Promise<Object> evaluateScriptFromURL(USVString url, boolean onlyForFrameScript);
+
+  Promise<Blob> capturePage(Rect source, Rect destination); // Works even for frozen pipelines
+
+  Promise<Blob> savePage(SaveType saveType);
+
+  Promise<Blob> downloadURL(USVString url);
+
+  Promise<Sequence<ContentBlocker>> getContentBlockers(ContentBlockerType type);
+
+  Printable asPrintable();
+  Editable asEditable();
+  Findable asFindable();
+  MultimediaManager asMultimediaManager();
+  HTTPObserverManager asHTTPObserverManager();
```

Comment:
Why can't those be just attributes `printManager`, `editManager`, `findManager`, `multimediaManager`, `networkManager` ?